serialization/rp_ser.c

```diff
@@ -0,0 +1,366 @@
+/*
+ * Copyright (c) 2020 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+#include <rp_ser.h>
+#include <rp_trans.h>
+#include <rp_os.h>
+#include <rp_errors.h>
+
+#define RP_LOG_MODULE SER_CORE
+#include <rp_log.h>
+
+#define RP_SER_RSP_INITIAL_ARRAY_SIZE 1
+#define RP_SER_CMD_EVT_INITIAL_ARRAY_SIZE 2
+
+static uint8_t endpoint_cnt;
+
+static cmd_handler_t cmd_handler_get(struct rp_ser *rp, uint8_t cmd)
+{
+	cmd_handler_t cmd_handler = NULL;
+
+	for (const struct rp_ser_cmd *iter = rp->decoders->cmd_begin;
+	     iter <  rp->decoders->cmd_end; iter++) {
+		if (cmd == iter->cmd) {
+			cmd_handler = iter->func;
+			break;
+		}
+	}
+
+	return cmd_handler;
+}
+
+static rp_err_t cmd_execute(struct rp_ser *rp, uint8_t cmd, CborValue *it)
+{
+	rp_err_t err;
+	cmd_handler_t cmd_handler;
+
+	cmd_handler = cmd_handler_get(rp, cmd);
+
+	if (cmd_handler) {
+		err = cmd_handler(it);
+	} else {
+		RP_LOG_ERR("Unsupported command received");
+		err = RP_ERROR_NOT_SUPPORTED;
+	}
+
+	return err;
+}
+
+static rp_err_t event_parse(struct rp_ser *rp, uint8_t evt, CborValue *it)
+{
+	rp_err_t err;
+	evt_handler_t evt_handler = NULL;
+
+	for (const struct rp_ser_evt *iter = rp->decoders->evt_begin;
+	     iter <  rp->decoders->evt_end; iter++) {
+		if (evt == iter->evt) {
+			evt_handler = iter->func;
+			break;
+		}
+	}
+
+	if (evt_handler) {
+		err = evt_handler(evt, it);
+	} else {
+		RP_LOG_ERR("Unsupported event received");
+		err = RP_ERROR_NOT_SUPPORTED;
+	}
+
+	return err;
+}
+
+static rp_err_t rp_ser_response_parse(struct rp_ser *rp, CborValue *it)
+{
+	rp_err_t err;
+
+	if (rp->rsp_handler) {
+		err = rp->rsp_handler(it);
+		if (err) {
+			return err;
+		}
+
+		rp->rsp_handler = NULL;
+	}
+
+	return rp_os_response_signal(rp);
+}
+
+static rp_err_t rp_ser_event_parse(struct rp_ser *rp, CborValue *it)
```

Comment:
This name is confusing with almost the same name `event_parse()` maybe call it event_id_parse() or perhaps event_process(). Or alternatively `event_parse()` could be renamed to `event_dispatch()`, since it effectively calls the registered decoder. Then rename cmd_parse to cmd_dispatch for consistency. Decide whether or not to use the _rp_ser_ prefix for static functions and stick to this decision. I would prefer to skip the prefix for simplicity.